### PR TITLE
fix of detach volume while OS is in boot state

### LINKED ISSUE #3290: KVM: if a volume is detached before OS has boot and migrated, admin not able to attach back
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Volumes, VM deployment
~~~

##### CLOUDSTACK VERSION

~~~
4.11
~~~

##### OS / ENVIRONMENT

Observed with KVM, should be agnostic of distro and version. We tested on:
```
# virsh version
Compiled against library: libvirt 4.5.0
Using library: libvirt 4.5.0
Using API: QEMU 4.5.0
Running hypervisor: QEMU 1.5.3
# cat /etc/centos-release
CentOS Linux release 7.4.1708 (Core) 
# uname  -a
Linux <stripped> 3.10.0-693.11.1.el7.x86_64 #1 SMP Mon Dec 4 23:52:40 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

##### SUMMARY

When a user creates a VM without starting it, and then start and detach the volume before the guest OS has booted, post migration the volume cannot be attached back. 

##### STEPS TO REPRODUCE

~~~
1. Create a VM with a data disk without starting it
2. Start the VM
3. Before the VM has booted, detach the disk
4. Migrate the data disk to another storage pool
5. Try to attach back the volume to the same VM
~~~

##### EXPECTED RESULTS

~~~
Volume being attached
~~~

##### ACTUAL RESULTS

~~~
Failed to attach volume test to VM VM-827b8faf-4ed6-414b-9c4f-29c7dcefd06c; org.libvirt.LibvirtException: XML error: target 'vdb' duplicated for disk sources '/mnt/610e0939-6665-311d-9401-bdcd354a275c/96912f59-95a5-458b-81c2-3e91c0687799' and '/mnt/610e0939-6665-311d-9401-bdcd354a275c/21d022d9-a2ba-4db0-acb4-e8f3044239c0'
~~~
It seems that the old record is not properly cleaned up when the volume is detached, and post migration we got a new path which creates a conflict and cloudstack is not able to attach the volume.

Workaround: 
This could be escaped if the user waits enough for the Guest OS to boot. After that detaching seems to be cleaning properly the records. 
Another way if the VM is already messed is to Stop and Start the VM, after that user is able to attach the volume.

### PR DESCRIPTION
**Pull request by slavkap**
#### Description

Fixes #3290

After detach device is executed we are checking into XML description that the device isn't there

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ X] Minor
- [ ] Trivial

### How Has This Been Tested?
Manual tests according to the  steps to reproduce

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
naking sure these ^^ are not a fluke
@blueorangutan test

**Comment by slavkap** [comment]
Hi @DaanHoogland, I think this will fail again, because when the tearDown() method is invoked it tries to detach the volume, but the OS is still in boot state. Maybe I have to add in the setUpClass to check with ssh connection that the VM has booted?

**Comment by DaanHoogland** [comment]
@slavkap sound good, makes sense that for such a fix the tests would need fixing as well

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@slavkap can you check build failure?

**Comment by slavkap** [comment]
@rhtyd, I'm actually not sure what to check. May I ask for some guidelines on what to check exactly?

**Review by rhtyd — APPROVED** [review]
LGTM did not test it

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java [hunk 1]
```diff
@@ -1166,6 +1167,14 @@ protected synchronized String attachOrDetachDevice(final Connect conn, final boo
...
                 s_logger.debug("Detaching device: " + xml);
                 dm.detachDevice(xml);
+                LibvirtDomainXMLParser parser = new LibvirtDomainXMLParser();
+                parser.parseDomainXML(dm.getXMLDesc(0));
+                List<DiskDef> disks = parser.getDisks();
+                for (DiskDef diskDef : disks) {
+                    if (StringUtils.contains(xml, diskDef.getDiskPath())) {
+                        throw new InternalErrorException("Could not detach volume. Probably the VM is in boot state at the moment");
+                    }
+                }
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java:1177]: code LGTM, for detach disk.

**Review by sureshanaparti — APPROVED** [review]

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java [hunk 2]
```diff
@@ -1166,6 +1167,14 @@ protected synchronized String attachOrDetachDevice(final Connect conn, final boo
             } else {
                 s_logger.debug("Detaching device: " + xml);
                 dm.detachDevice(xml);
+                LibvirtDomainXMLParser parser = new LibvirtDomainXMLParser();
+                parser.parseDomainXML(dm.getXMLDesc(0));
+                List<DiskDef> disks = parser.getDisks();
+                for (DiskDef diskDef : disks) {
+                    if (StringUtils.contains(xml, diskDef.getDiskPath())) {
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java:1174]: @slavkap can you check if this is applicable for attach disk as well (when VM is either booting or shutting down)?

Comment (slavkap) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java:1174]: Sure, @sureshanaparti, I'll do the test and will let you know.

**Comment by rhtyd** [comment]
@blueorangutan test

**Review by rhtyd — APPROVED** [review]

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java (hunk 2 above)
Comment (slavkap) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java:1174]: @sureshanaparti, sorry for the late response, but I didn't have time to check it yesterday! The change is applied only to detach a volume when the VM is in Running or Stopped state

**Pull request merged**